libpyvinyl/Parameters/Parameter.py

```diff
@@ -8,126 +8,168 @@ class Parameter(AbstractBaseClass):
 
         This could be expanded to raise an exception, or such could be in is_legal
         """
         if self.is_legal(value):
-            self.value = value
+            self._value = value
         else:
-            print("WARNING: Value of parameter '" + self.name
-                  + "' illegal, ignored.")
+            raise ValueError("Value of parameter '" + self.name + "' illegal.")
 
-    def is_legal(self, value=None):
+    def is_legal(self, values=None):
         """
         Checks whether or not given or contained value is legal given constraints.
 
         Illegal intervals have the highest priority to be checked. Then it will check the
         legal intervals and options. The overlaps among the constrains will be overridden by
         the constrain of higher priority.
         """
-        if value is None:
-            value = self.value
-
-        # Check illegal intervals
-        for illegal_interval in self.illegal_intervals:
-            if illegal_interval[0] < value < illegal_interval[1]:
-                return False
-
-        # Check legal intervals
-        is_inside_a_legal_interval = False
-        for legal_interval in self.legal_intervals:
-            if legal_interval[0] < value < legal_interval[1]:
-                is_inside_a_legal_interval = True
 
-        if not is_inside_a_legal_interval and len(self.legal_intervals) > 0:
-            return False
+        truths = []
+        if values is None:
+            values = self._value
 
-        # checked intervals, can return if options not used (frequent case)
-        if len(self.options) == 0:
-            return True
+        if not hasattr(values, "__iter__") or isinstance(values, str):
+            #            print(str(hasattr(values, "__iter__")) + str(values))
+            value = values
 
-        for option in self.options:
-            if option == value:
-                # If the value matches any option, it is legal
-                return True
+            for option in self.options:
+                if option == value:
+                    #
+                    return self._options_are_legal
+
+            # Check illegal intervals
+            for interval in self.intervals:
+                if interval[0] < value < interval[1]:
+                    #
+                    return self._intervals_are_legal
+            # T: are legal values
+            # F: are illegal values
+            # 0: empty list
+            #
+            # P I -> truth
+            # T T -> F
+            # F T -> F
+            # 0 T -> F
+            # T F -> T
+            # F F -> T
+            # 0 F -> T
+            # T 0 -> F
+            # F 0 -> T
+            # 0 0 -> T
+
+            return (not self._intervals_are_legal and len(self.intervals) > 0) or (
+                len(self.intervals) == 0
+                and (not self._options_are_legal or len(self.options) == 0)
```

Comment:
This is not easy to read.

I suggest we create two additional methods that is_legal can call. is_value_legal_intervals and is_value_legal_options.

is_value_legal_intervals:
```
for interval in self._intervals:
    if interval[0] <= value <= interval[1]
        return self._intervals_are_legal

# here because value was in no intervals
if len(self._intervals) > 0:
    return not self._intervals_are_legal

return True # if no restrictions defined, true
```

is_value_legal_options:
```
for option in self._options:
    if option == value:
        return self._options_are_legal

# here because value did not match any options
if len(self._options) > 0:
    return not self._options_are_legal

return True # if no restrictions defined, legal
```

Then is_legal (for a single value) can be written:
```
return self.is_value_legal_intervals(value) and self.is_value_legal_options(value)
```

Reply:
I got your suggestion, but the truth table is not the same...

Reply:
I can split the last return. Not super efficient, but still ok and maybe a bit more readable. Here's my proposal:
```
            # obvious, if no conditions are defined, the value is always legal
            if len(self._options) == 0 and len(self._intervals) == 0:
                return True

            # first check if the value is in any defined discrete value
            for option in self._options:
                if option == value:
                    return self._options_are_legal

            # secondly check if it is in any defined interval
            for interval in self._intervals:
                if interval[0] <= value <= interval[1]:
                    return self._intervals_are_legal

            # at this point the value has not been found in any interval
            # if intervals where defined and were forbidden intervals, the value should be accepted
            if len(self._intervals) > 0:
                return not self._intervals_are_legal

            # if there where no intervals defined, then it depends if the discrete values were forbidden or allowed
            return not self._options_are_legal
```

Reply:
That is much easier to read, thanks for putting in the effort to clarify!